src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -763,6 +763,78 @@ private void DecodeDllImportAttribute(ref DecodeWellKnownAttributeArguments<Attr
                     preserveSig);
             }
         }
+
+        private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
+        {
+            Debug.Assert(arguments.AttributeSyntaxOpt is object);
+
+            if (MethodKind != MethodKind.Ordinary)
+            {
+                // Ignore cases where there will already be a warning due to the AttributeTargets.Method usage
+                // restriction on the attribute definition in the framework.
+                if (MethodKind != MethodKind.Constructor && MethodKind != MethodKind.StaticConstructor)
+                {
+                    arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerMethodMustBeOrdinary, arguments.AttributeSyntaxOpt.Location);
+                }
+                return;
+            }
+
+            Debug.Assert(ContainingType is object);
+
+            if (isInaccessible(this) || containingTypes(this).Any(isInaccessible))
+            {
+                var topLevelType = containingTypes(this).Last();
```

Comment:
>var topLevelType = containingTypes(this).Last(); [](start = 16, length = 48)

I do not believe it is worth calculating this and including into the error message